### DISCUSSION TITLE
[WS2122] Homework 2 Discussion

### DISCUSSION THREAD

**Discussion by changkun** [mimuc/gp-ws2021#33] [Discussion] [answered]
This thread is open for the discussion of homework 2.

**Comment by nico778** [comment]
Hi, I tried to use my halfedge-structure from the first homework, but nothing is displayed after importing a mesh. I'm guessing my halfedge-structure must be wrong then, or do we have to change something else?

**Reply by changkun** [reply]
Sorry for the late response. I think we are not that strict here. So the results are evaluated on the rendering result, if you could produce visually comparable results, then they are the same.

From the second exercise, the change might be from an open boundary mesh to a watertight mesh (although I have not yet evaluated all submissions yet)

**Comment by Consti10** [comment]
Hello,
Quick question regarding the "fork and branch workflow" should the pull request for the solution of the second homework also
contain the previous homework solution or not ? I'd think the files on the second pull request should look like this,right ?

**Option 1(yes):**
homeworks/constantin/2-ddg
**And not like this:**
homeworks/constantin/1-halfedge
homeworks/constantin/2-ddg

**Reply by changkun** [reply]
The new PR should not contain previous commits. Hence, yes, option 1 is preferred.

**Comment by Consti10** [comment]
Are we allowed to extend the geometric math functions of the project ?
For examle a helper to construct a Vector from 2 Vertices. 
`return new Vector(pos2.x-pos1.x,pos2.y-pos1.y,pos2.z-pos1.z);`
aka I'd like to not write the above code every time ;)

**Reply by Consti10** [reply]
Found the proper helper(s) in "vec.ts".

**Comment by JohannesMerkt** [comment]
Hi,
im stuck. So far all i have done was to take the halfedge code from the first homework and copied it over while making sure it works in the first assignment, but i cant see anything at the moment. What do i have to do to see my geometry? Before i start working on the normal and curvature implementations

**Reply by JohannesMerkt** [reply]
Got it so if you just want to see the geometry try setting the normal vector to 1,1,1,1 and the curvature to 1 too

**Reply by Consti10** [reply]
Yes, then you should get an image.

**Reply by changkun** [reply]
Yes, that's great sharing. I think we briefly talked about this in precense, did we?

**Reply by Consti10** [reply]
yes, but we only found out that you have to set both normal and curvature to default values late after the lecture :)

**Comment by JohannesMerkt** [comment]
Hi,
I am confused. The halfedge implementation has a function called angle with a todo message: compute the tip angle at this edge. So does that mean I should calculate the angle of the corner of the triangle or should I calculate the redefined angle which is considering  the geometry is smooth?
![grafik](https://user-images.githubusercontent.com/25324640/141157566-d5387c45-e043-4c42-bc34-0e517ebccd77.png)
I assume just calculating the triangle angle makes the most sense here

**Reply by Consti10** [reply]
I think the `angle()` method is intended to be used when calculating the Normals in "AngleWeighted" mode. Therefore I'd calculate the "corner Angle" aka Theta1,2,... .

**Reply by Consti10** [reply]
Actually, I think I only now understand your question correctly - basically you are asking if in angle() we should scale the angle such that the sum of angles is 2Pi again ? No idea ;)

**Reply by JohannesMerkt** [reply]
Now that i finished it. Im quite sure its without having the sum of the angles sum up to 2Pi

**Reply by changkun** [reply]
Actually `angle` is the angle of the opposite side of an halfedge.

**Reply by changkun** [reply]
Or can be interpreted as the adjacency angle of an halfedge (depends on the implementation)

**Comment by alvaxgln** [comment]
Hey i'm having a lot of problems...

My halfedge implementation worked fine for homework one, so first i copied over halfedge.ts, but that gave me an error cause it couldn't find "linalg.mat" and "linalg/quaternion" (are those needed? Should i copy those over too?)
So i just replaced the buildMesh() function with my implementation from homework one, and set the normal and curvature to 1.
But when i start the program it just gives me a black screen, with the controls, and none of the controls are clickable.

Not sure what the problem is here

**Reply by Consti10** [reply]
The default mesh now has much more vertices than the previous one. In my case, the halfedge building code was not using a hash map an therefore took ~20seconds to build. So I thought at first the controlls were not working.

**Reply by changkun** [reply]
> couldn't find "linalg.mat" and "linalg/quaternion" 

These are actually re-located. So you may have to change the code a little bit rather than a direct use. The purpose is to let you rethink the whole situation and do some adaption.

I also uploaded a reference solution, which you may use directly to get faster load speed.

**Reply by alvaxgln** [reply]
Ah, ok thank you

**Reply by JohannesMerkt** [reply]
If nothing is happening and your browser tab freezes its likely that your code isnt ver optimized. With my code from the first homework it took minutes to load the mesh 😂

**Comment by Consti10** [comment]
When calculating the "Gaussian Curvature" aka `CurvatureMethod.Gaussian` can we just use the angle defect ?
![präsi_gpp_screenshot](https://user-images.githubusercontent.com/16462285/141294459-4d198a9c-8f68-4e9c-a352-8711c5d2fc64.JPG)

**Reply by changkun** [reply]
Yes :)

**Comment by Consti10** [comment]
Should you be able to observe a difference on the normals when changing the "normalMethod" ?
In my case, all 3 of them give exactly the same results, at least when looking at the rendered Image. Haven't checked the raw vec values yet.

**Reply by JohannesMerkt** [reply]
At first glance I thought they where the same but on close inspection you can see some minor changes. Btw you can zoom an pan on the 3d model

**Reply by changkun** [reply]
You should be able to observe changes between different normal methods, if you activate showing normals.

Note that you must be very close to the mesh to see the differences.

**Reply by Consti10** [reply]
okay, I will look more closely on the rendered image. Looking at the raw vec values, they are slightly different.

**Reply by Consti10** [reply]
Thanks, now I can see it too. Most noticable at the ears ;)

**Comment by alvaxgln** [comment]
I'm currently trying to calculate the area of the voronoi cell around a vertex.
Is there a more convenient way to do that, than finding the circumcenter of each face?

**Reply by nico778** [reply]
if I'm not mistaken it can be done with the cotangents of the halfedges

**Comment by changkun** [comment, marked as answer]
Close due to the end of the submission period.
